zulip_updates: Send zulip updates based on `zulip_update_announcements_level`.

prakhar1144 opened the pull request:
Fixes #28604 

<details><summary>Settings</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/b5f287d2-e7d9-42dd-9ef5-f693aefec8a3">
</img>
</details> 

<details><summary>Onboarding message when new realm is created</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/d5936294-a97d-4055-b436-cf85c8398d22">
</img>
</details> 

<details><summary>Old realm with zulip update stream configured</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/91202488-d667-4962-9418-2111c2f9500a">
</img>
</details> 

<details><summary>Old realm with zulip update stream not configured</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/e068751e-3d46-4296-92e8-b0aac60341c7">
</img>
</details> 

---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

prakhar1144 commented:
This is functionally ready, apart from couple of small questions I'll post shortly in #design

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -4521,6 +4521,18 @@ paths:
+                                      type: integer
+                                      description: |
+                                        The ID of the stream to which automated messages announcing
+                                        new features, new keyboard shortcuts, etc are sent.
```
"new features or other end-user updates about the Zulip software." is probably a better description.

prakhar1144 replied:
Updated.

timabbott commented on zerver/actions/create_realm.py:
```diff
@@ -295,11 +296,17 @@ def do_create_realm(
 
+    # New realm is initialized with the latest zulip updates level
+    # as it shouldn't receive a bunch of old updates.
+    # Should we send a DM to admins, if they want to change the stream for updates?
```
I think we don't need DMs here, but we could consider sending including in our onboarding notices a message to a topic in the default value of this stream for new organizations; @alya FYI as something to consider in your onboarding project but we can probably keep it out of this PR in any case.

timabbott replied:
Oh looks like you already have one in `onboarding.py`, so yeah I think you can delete this comment.

prakhar1144 replied:
Yeah, I just forgot to remove that comment. Sorry for the confusion.

timabbott commented on zerver/lib/onboarding.py:
```diff
@@ -314,6 +314,18 @@ def send_initial_realm_messages(realm: Realm) -> None:
+                "[turn this feature off](/#organization/organization-settings) altogether."
+            )
+        ).format(
+            move_content_help_url="/help/move-content-to-another-stream",
```
Let's use variables, `settings_url` and `settings_disable_url` here, so we don't have to ask translators to change anything if the URLs change. (Right now, those URLs will be the same, but we might in the future be able to add a vertical-position-within-page anchor at some point)

prakhar1144 replied:
Makes sense. Made the changes.

timabbott commented on zerver/lib/import_realm.py:
```diff
@@ -957,6 +957,9 @@ def do_import_realm(import_dir: Path, subdomain: str, processes: int = 1) -> Rea
     re_map_foreign_keys(data, "zerver_realm", "signup_announcements_stream", related_table="stream")
+    re_map_foreign_keys(
+        data, "zerver_realm", "zulip_update_announcements_stream", related_table="stream"
+    )
```
So this probably is something we should fork to its own follow-up issue, but when an organization gets imported from another product, do any of these announcement stream fields get set? I assume the answer is no.

We probably want to initialize them as we would for a new realm somehow when doing data imports from non-Zulip servers; probably a little prep PR we could do for that is make sure the export tarballs have a clear "what product was this exported from" field -- and then we can build on top of that.

prakhar1144 replied:
> when an organization gets imported from another product, do any of these announcement stream fields get set?

No, they aren't.


#29041

timabbott commented on zerver/tests/test_home.py:
```diff
@@ -465,7 +466,7 @@ def test_num_queries_with_streams(self) -> None:
         # Then for the second page load, measure the number of queries.
-        with self.assert_database_query_count(44):
+        with self.assert_database_query_count(45):
             result = self._get_home_page()
```
These database queries are starting to add up a bit; we should look at whether there's a refactor we can/should do.

Not a blocker for this PR; I'll open an issue.

prakhar1144 replied:
#31028

timabbott commented on zerver/management/commands/send_zulip_updates.py:
```diff
@@ -0,0 +1,134 @@
+                    do_send_messages(messages)
+                if new_zulip_updates_level:
+                    realm.zulip_updates_level = new_zulip_updates_level
+                    realm.save(update_fields=["zulip_updates_level"])
```
Let's put most of this code in one or more functions in a new `zerver/lib/zulip_update_announcements.py` and name this file `send_zulip_update_announcements.py`. A few reasons for that proposal:
* We should use the same term for the concept everywhere, rather than having small variations in how something is referenced.
* `lib/` files are easier to unit-test than code sitting in management commands; so we try to keep the management commands small/simple, pointing to library code that does the actual work.

prakhar1144 replied:
* Moved everything to `zerver/lib/zulip_update_announcements.py`.
* Renamed the file to `send_zulip_update_announcements.py`

TODO: Add unit tests.

timabbott commented on zerver/management/commands/send_zulip_updates.py:
```diff
@@ -0,0 +1,134 @@
+            Q(zulip_updates_level__isnull=True)
+            | Q(zulip_updates_level__lt=latest_zulip_updates_level),
+            deactivated=False,
+        )
```
Let's add a comment noting why we're doing `deactivated=False` here -- if the realm is later reactivated, then we'll actually send the notices that it missed while it was inactive. I think that's correct behavior.

prakhar1144 replied:
Thought about this, Should we send bunch of notices when someone reactivate ? -- If the time gap between deactivation - reactivation is large (say ~1-2 year) that will cause a lot of notices...

I think, the behaviour here will depend on "what are the practical cases when someone reactivate their realm"

alya replied:
I think we are OK with a bundle of notices. I expect us to send these like once a month, not once a day, so it really won't be that much.

prakhar1144 replied:
Comment added.
```
    # Filter out deactivated realms. When a realm is later reactivated,
    # send the notices it missed while it was deactivated.
```

timabbott commented on zerver/management/commands/send_zulip_updates.py:
```diff
@@ -0,0 +1,134 @@
+                        event_type=RealmAuditLog.REALM_PROPERTY_CHANGED,
+                        extra_data={
+                            RealmAuditLog.OLD_VALUE: None,
+                            RealmAuditLog.NEW_VALUE: 1,
```
These values seem wrong, since we were already at 1?

prakhar1144 replied:
`zerver/lib/zulip_update_announcements.py` has the updated code with this logic inside `is_group_direct_message_sent_to_admins_atleast_one_week_ago`

timabbott commented on zerver/management/commands/send_zulip_updates.py:
```diff
@@ -0,0 +1,134 @@
+                if messages:
+                    do_send_messages(messages)
+                if new_zulip_updates_level:
+                    realm.zulip_updates_level = new_zulip_updates_level
```
We probably should write audit log entries for the change to `zulip_updates_level` here? And possibly those audit log entries should have a bonus field for the `message_id` of the message that was sent.

timabbott commented on zerver/management/commands/send_zulip_updates.py:
```diff
@@ -0,0 +1,134 @@
+                        direct_message = internal_prep_private_message(
+                            sender=sender, recipient_user=administrator, content=content
+                        )
+                        messages.append(direct_message)
```
I think the idea we'd had in mind was a group DM to all of the administrators, so that they know exactly who received it and can easily coordinate by replacing there. Worth testing that replying to such a thread including Notification Bot works nicely, but I think it should -- and then we'd only have a single message ID that we can put in the audit log entry.

prakhar1144 replied:
Makes sense. Changed it to group DM.

I have used the term `group_direct_message` instead of `huddle`, (because, I think we're renaming huddle to group_direct_message -- I think there was a discussion on CZO ?) -- If I'm not correct, I will change it back to huddle.

timabbott replied:
Yeah that sounds reasonable.

prakhar1144 commented:
Addressed the review comments & replied back.

TODO:
* Add unit tests.
* https://github.com/zulip/zulip/pull/28720#discussion_r1486598760

Questions:
* https://github.com/zulip/zulip/pull/28720#discussion_r1488205661
* https://github.com/zulip/zulip/pull/28720#discussion_r1488213734
* Where should we put the markdown file where the zulip updates are enlisted ?

prakhar1144 commented:
Updates in recent push:

* Added unit test
* Addressed review comments

Question:
* Where should we put the markdown file where the zulip updates are enlisted ?

timabbott commented on zerver/actions/message_send.py:
```diff
@@ -1974,17 +1974,25 @@ def internal_send_stream_message_by_name(
+
     if message is None:
         return None
+
```
This seems like a fine refactor but usually I'd do it in a prep commit.

prakhar1144 replied:
Yeah, that's correct. I missed that. Done.

timabbott commented on zerver/lib/zulip_update_announcements.py:
```diff
@@ -0,0 +1,177 @@
+        content = zulip_updates.read()
+        match = re.search(r"\*\*Zulip update announcements level (\d+)\*\*", content)
+        assert match is not None
+        return int(match.group(1))
```
So I think @alya's proposal is that we should just configure them as `"""` sections in this Python file, rather than having a specially formatted Markdown file. The motivation for that is that if we ever want to make some of them conditional on anything (say, organization type), that'll work.

timabbott commented on zerver/lib/zulip_update_announcements.py:
```diff
@@ -0,0 +1,177 @@
+    current_level: int, latest_level: int, sender: UserProfile, realm: Realm
+) -> List[Optional[SendMessageRequest]]:
+    messages = []
+    topic_name = str(realm.ZULIP_UPDATE_ANNOUNCEMENTS_TOPIC_NAME)
```
Does this need to be inside a `with override_language(realm.default_language)` block? I think that's a thing we do elsewhere for `gettext_lazy` tagged string fields like this when running code in a cron job (and thus not preferring to use Django's built-in borrowing of the user's language).

prakhar1144 replied:
Yeah, that's right.

Added `with override_language(realm.default_language)` in `internal_prep_group_direct_message_for_old_realm` too.

I was wondering, why don't we use translation for topic & stream name of welcome messages. Is it intended ?

timabbott replied:
Good question; that may be a decision dating to when we didn't ask for an organization language before creating the organization, and so it would have been pointless to use any language as the value we'd get was always English. Worth discussing in chat.zulip.org.

timabbott commented on zerver/lib/zulip_update_announcements.py:
```diff
@@ -0,0 +1,177 @@
+    assert stream is not None
+    while current_level < latest_level:
+        content = get_zulip_update_announcements_message_for_level(level=current_level + 1)
+        messages.append(
```
Call this variable `message_requests`, for readability.

prakhar1144 replied:
Done.

timabbott commented on zerver/lib/zulip_update_announcements.py:
```diff
@@ -0,0 +1,177 @@
+
+        messages = []
+        new_zulip_update_announcements_level = None
+        with transaction.atomic():
```
I think you'll need to change the `with transaction.atomic():` to use `savepoint=False` -- we never want to use savepoints, and I noticed that it doesn't have that marker on it.

prakhar1144 replied:
Done.

prakhar1144 commented on zerver/lib/zulip_update_announcements.py:
```diff
@@ -0,0 +1,177 @@
+    realm: Realm, sender: UserProfile
+) -> Optional[SendMessageRequest]:
+    administrators = realm.get_human_admin_users()
+    emails = [administrator.email for administrator in administrators]
```
Since, we already have `user_profiles`, so included `Addressee.for_user_profiles` to use.

Using `Addressee.for_user_ids` will lead to "fetch userprofiles/administrators"  -> "pass user IDs" -> "again fetch user profiles".

prakhar1144 commented:
Addressed the review comments.

This is ready, apart from the "What group DM should we send to admins" question I [posted in CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/New.20feature.20notification.20system.20.2328604/near/1744086)

timabbott commented on zerver/lib/zulip_update_announcements.py:
```diff
@@ -0,0 +1,207 @@
+zulip_update_announcements: List[ZulipUpdateAnnouncement] = [
+    ZulipUpdateAnnouncement(
+        level=1,
+        message="",
```
Yeah, it feels a bit odd. I guess another theory would be that we shouldn't change `zulip_update_announcements_level` off `None` when sending the initial group DM, and instead only use `RealmAuditLog` to keep track of that state (Checking whether it's been sent at all, in addition to checking whether it's been at least a week).

timabbott replied:
OK that works for me.

timabbott commented on zerver/lib/zulip_update_announcements.py:
```diff
@@ -0,0 +1,207 @@
+    with override_language(realm.default_language):
+        topic_name = str(realm.ZULIP_UPDATE_ANNOUNCEMENTS_TOPIC_NAME)
+        while current_level < latest_level:
+            content = _(get_zulip_update_announcements_message_for_level(level=current_level + 1))
```
What is this `_()` wrapping supposed to do? Internationalization doesn't work that way; see 
https://zulip.readthedocs.io/en/latest/translating/internationalization.html, but in short, `_()` needs to always wrap an actuals tring for it to be tagged for translation, because the discovery code actually parses the Python looking for specific patterns.

I worry a bit that it won't make sense to tag these updates for translation, because they are quite unlikely to be translated during the time between when we draft them and when they are published. So it's possible we should just not be translating the topic either...

timabbott replied:
Chatted with @alya and she agrees probably we shouldn't attempt to translate the updates themselves, but the topic should be translated as currently implemented. 

So I think you can just delete the `_()` here and leave a comment in the `zulip_update_announcements` definition.

prakhar1144 replied:
> because they are quite unlikely to be translated during the time between when we draft them and when they are published.
> but the topic should be translated as currently implemented.

Are we sure about translating "topic names" before we make it available for cloud customers ?

prakhar1144 replied:
Made the changes to translate only the "topic name". The updates itself are not translated.

timabbott commented on zerver/lib/zulip_update_announcements.py:
```diff
@@ -0,0 +1,207 @@
+
+        messages = []
+        new_zulip_update_announcements_level = None
+        with transaction.atomic(savepoint=False):
```
Oops, I realize I didn't say the thing thing -- it's inner the `transaction.atomic` in `do_send_messages` that needs this parameter, not this one. (Basically a transaction-inside-a-transaction makes a savepoint by default).

You can change it in a prep commit.

prakhar1144 replied:
Done.

timabbott commented on zerver/lib/onboarding.py:
```diff
@@ -314,6 +314,19 @@ def send_initial_realm_messages(realm: Realm) -> None:
+                "If you like, you can [configure]({settings_url}) a different stream "
+                "for these updates (and [move]({move_content_help_url}) the initial updates there), "
+                "or [turn this feature off]({settings_url}) altogether."
+            )
```
@prakhar1144 so I think this message might be what we want as "level 1" in the system -- that way, existing installations that upgrade to Zulip 9.0 will get this message explaining what the topic is about. We still want it to be also send during realm creation, such that the first message in the "Zulip updates" topics should always be this one explaining what the topic is about.

I guess we also want it to be sent for new organizations, though; perhaps you can figure out a way to not have the content be duplicated?

And then probably @alya will want to rewrite the content to be designed for both of those contexts.

prakhar1144 replied:
> I guess we also want it to be sent for new organizations

This was our plan as discussed on CZO:
> New realms are initialized with the current/latest zulip_updates_level -- i.e. so newly created realms won't get a bunch of old notices (though we may want to add an onboarding message sent to the default stream location for this).

So, I think, new realms should only get the onboarding message & then new updates onwards (no duplication). Right? -- or we want to send all the notices ?

prakhar1144 replied:
New realms created get the onboarding message.

Existing realms get the same message as a part of zulip update announcements.

New realms will be initialized to the latest realm, so they won't get the first announcement -- i.e. no duplication.

timabbott replied:
> Existing realms get the same message as a part of zulip update announcements.

I think they only get it in a direct message to administrators, though -- we probably want something that goes to the actual topic.

timabbott replied:
Ahh, OK, I see you have the same text in both places. I think basically this is ready for me to do a wording edit pass with @alya and then we'll probably be good to go.

prakhar1144 commented:
Ready to review.

Addressed the review comments.

alya commented on zerver/lib/zulip_update_announcements.py:
```diff
@@ -0,0 +1,206 @@
+        ).format(organization_settings_url="/#organization/organization-settings")
+    else:
+        content = (
+            "Users in your organization will now receive updates about new features of Zulip "
```
This should be "about new Zulip features", like the other messages.

alya commented on zerver/lib/zulip_update_announcements.py:
```diff
@@ -0,0 +1,206 @@
+        message=(
+            "This topic will receive automated updates about new Zulip features.\n"
+            "\n"
+            "If you like, you can [configure]({settings_url}) a different stream "
```
I'm thinking this should be tweaked, so that it makes sense for both admins and non-admins. Let's try changing "If you like, you can" -> "Organization administrators can".

alya commented on zerver/lib/zulip_update_announcements.py:
```diff
@@ -0,0 +1,206 @@
+            "You can now [configure]({organization_settings_url}) a stream where Zulip "
+            "will send updates to users in your organization about new Zulip features. "
+            "These notifications are currently turned off in your organization."
+        ).format(organization_settings_url="/#organization/organization-settings")
```
This is a bit had to read. Let's try "will send updates to users in your organization" -> "will send updates".

alya commented:
I posted some wording tweaks above. I think we should also write up a help center article, and link to it from these messages.

prakhar1144 commented:
<details><summary>Group DM with announcement stream not None</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/18fca519-6a10-4705-bb3f-92f4facb6ca6">
</img>
</details> 

<details><summary>Group DM with announcement stream None</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/3a412ff4-f9d0-4148-a3eb-33a68e2cbdef">
</img>
</details> 

<details><summary>First update</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/00840faf-8f3c-4e06-835a-10c39205cae0">
</img>
</details> 

<details><summary>Onboarding message</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/035ab8e2-7b1a-4a3d-b9f1-e89a362a1456">
</img>
</details>

timabbott commented on zerver/lib/zulip_update_announcements.py:
```diff
@@ -0,0 +1,212 @@
+            "Organization administrators can [configure]({settings_url}) a different stream "
+            "for these updates (and [move]({move_content_help_url}) the initial updates there), "
+            "or [turn this feature off]({settings_url}) altogether."
+        ).format(
```
Can we redo all these strings using the `"""` Python quote blocks? It'll be a lot more convenient to edit in that form.

prakhar1144 replied:
Done.

prakhar1144 commented:
Addressed the [review comment](https://github.com/zulip/zulip/pull/28720#discussion_r1513593054).

alya commented on zerver/lib/zulip_update_announcements.py:
```diff
@@ -0,0 +1,212 @@
+        message="""\
+This topic will receive [automated updates]({zulip_update_announcements_help_url}) \
+about new Zulip features.
+
```
@timabbott and I chatted about this one, and are now thinking the org admin part can be skipped here. Maybe something like this?

To help you learn about improvements in the Zulip app, this topic will receive [automated updates]({zulip_update_announcements_help_url}) about new Zulip features. [Learn more](/help/configure-notification-bot#zulip-update-announcements).

alya replied:
It might be worth making the help center link more explicit like this (and fleshing out the help center section more).

@timabbott what do you think?

alya replied:
https://zulip.com/help/configure-notification-bot#zulip-update-announcements is the section to link to now.

We still need to migrate this help page to a new URL (/help/configure-automated-notices). @prakhar1144 maybe you could take care of that?

alya replied:
Maybe something like this for the message text?

---

Zulip is introducing **Zulip updates**! To help you learn about new features and configuration options, this topic will receive messages about important changes in Zulip.

You can read update messages whenever it's convenient, or [mute](https://zulip.com/help/mute-a-topic) this topic if you are not interested. If you organization does not want to receive these announcements, they can be disabled. [Learn more](https://github.com/zulip/zulip/pull/%7Bzulip_update_announcements_help_url%7D).

---

timabbott replied:
Yeah that works for me!

prakhar1144 replied:
<details><summary>Updated message</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/7af1bb72-3f5a-45b7-9c35-4024ac00db6d">
</img>
</details>

prakhar1144 commented:
Updates in the recent push:
* Addressed https://github.com/zulip/zulip/pull/28720#discussion_r1531017569 in the first commit.
* Addressed https://github.com/zulip/zulip/pull/28720#discussion_r1531031957

One question for Tim (Not a blocker for this PR):
* If we migrate a help page to a new URL (like we are doing in this PR) -- won't it affect the URLs indexed by search engine?

timabbott commented:
> If we migrate a help page to a new URL (like we are doing in this PR) -- won't it affect the URLs indexed by search engine?

We always do a redirect when we rename a /help/ article; see `zerver/lib/url_redirects.py`.

timabbott commented on zerver/lib/onboarding.py:
```diff
@@ -314,6 +314,23 @@ def send_initial_realm_messages(realm: Realm) -> None:
+            + _(
+                "You can read update messages whenever it's convenient, or [mute]({mute_topic_help_url}) "
+                "this topic if you are not interested. If your organization does not want to receive "
+                "these announcements, they can be disabled. [Learn more]({zulip_update_announcements_help_url})."
```
Made a tweak to this and posted all 3 messages in #design for final feedback, but I think this version is passable and we can merge and deploy to chat.zulip.org as-is.

timabbott commented:
Along with adding the `url_redirects.py` redirect, I made the following wording tweaks:

```diff
diff --git a/zerver/lib/onboarding.py b/zerver/lib/onboarding.py
index e9a40d0898..a22c710170 100644
--- a/zerver/lib/onboarding.py
+++ b/zerver/lib/onboarding.py
@@ -315,17 +315,15 @@ def send_initial_realm_messages(realm: Realm) -> None:
         )
 
         content_of_zulip_update_announcements_topic_name = (
-            _(
-                "Zulip is introducing **Zulip updates**! To help you learn about new features and "
-                "configuration options, this topic will receive messages about important changes in Zulip."
-            )
-            + "\n"
-            "\n"
-            + _(
-                "You can read update messages whenever it's convenient, or [mute]({mute_topic_help_url}) "
-                "this topic if you are not interested. If your organization does not want to receive "
-                "these announcements, they can be disabled. [Learn more]({zulip_update_announcements_help_url})."
-            )
+            _("""
+Welcome! To help you learn about new features and configuration options, \
+this topic will receive messages about important changes in Zulip.
+
+You can read these update messages whenever it's convenient, or \
+[mute]({mute_topic_help_url}) this topic if you are not interested. \
+If your organization does not want to receive these announcements, \
+they can be disabled. [Learn more]({zulip_update_announcements_help_url}).
+            """)
         ).format(
             zulip_update_announcements_help_url="/help/configure-automated-notices#zulip-update-announcements",
             mute_topic_help_url="/help/mute-a-topic",
diff --git a/zerver/lib/zulip_update_announcements.py b/zerver/lib/zulip_update_announcements.py
index 17658f62fb..94df79fa59 100644
--- a/zerver/lib/zulip_update_announcements.py
+++ b/zerver/lib/zulip_update_announcements.py
@@ -34,7 +34,7 @@ zulip_update_announcements: List[ZulipUpdateAnnouncement] = [
 Zulip is introducing **Zulip updates**! To help you learn about new features and \
 configuration options, this topic will receive messages about important changes in Zulip.
 
-You can read update messages whenever it's convenient, or [mute]({mute_topic_help_url}) \
+You can read these update messages whenever it's convenient, or [mute]({mute_topic_help_url}) \
 this topic if you are not interested. If your organization does not want to receive these \
 announcements, they can be disabled. [Learn more]({zulip_update_announcements_help_url}).
 """.format(
```

@alya FYI, the `\` at end-of-lines is necessary for line-wrapping; may be a bit fiddly, and we can probably write a function to just remove single-newlines in these notices to avoid the need for them; @prakhar1144 that might be a nice thing to do soon as preparation for the project to rewrite our existing onboarding notices, which I expect you'll be driving.

timabbott commented:
Merging without CI passing, as CI isn't running thanks to a Launchpad outage for a few hours, and it's not clear that will end.

Huge thanks for all the work on this @prakhar1144! I believe the other follow-up if we need to add a cron job to run the new management command. See `puppet/zulip/manifests/app_frontend_once.pp` for where we do that.
